dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DiagnosticsHelper.java

```diff
@@ -0,0 +1,387 @@
+import io.dockstore.webservice.core.User;
+import io.dropwizard.core.setup.Environment;
+import jakarta.ws.rs.container.ContainerRequestContext;
+import jakarta.ws.rs.container.ContainerRequestFilter;
+import jakarta.ws.rs.container.ResourceInfo;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.ext.Provider;
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryPoolMXBean;
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.glassfish.jersey.server.ContainerRequest;
+import org.glassfish.jersey.server.ContainerResponse;
+import org.glassfish.jersey.server.monitoring.ApplicationEvent;
+import org.glassfish.jersey.server.monitoring.ApplicationEventListener;
+import org.glassfish.jersey.server.monitoring.RequestEvent;
+import org.glassfish.jersey.server.monitoring.RequestEventListener;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.stat.SessionStatistics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DiagnosticsHelper {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DiagnosticsHelper.class);
+    private static final double MILLISECONDS_PER_SECOND = 1e3;
+    private static final double NANOSECONDS_PER_SECOND = 1e9;
+    private static final double BYTES_PER_MEGABYTE = 1e6;
+    private static final String RESOURCE_METHOD_PROPERTY_NAME = "io.dockstore.webservice.helpers.DiagnosticsHelper.resourceMethod";
+    private static final int SEVEN = 7;
+
+    private Environment environment;
+    private SessionFactory sessionFactory;
+    private DockstoreWebserviceConfiguration.DiagnosticsConfig config;
+
+    private Logger logger;
+    private CensorHelper censorHelper;
+    private MemoryMXBean memoryMXBean;
+    private ThreadMXBean threadMXBean;
+
+    public DiagnosticsHelper() {
+        this(LOG);
+    }
+
+    public DiagnosticsHelper(Logger logger) {
+        this.logger = logger;
+        this.censorHelper = new CensorHelper(readFrequencies());
+        // Initialize the beans that we know are singletons, just in case there are performance issues with repeatedly retrieving them.
+        this.memoryMXBean = ManagementFactory.getMemoryMXBean();
+        this.threadMXBean = ManagementFactory.getThreadMXBean();
+    }
+
+    public void start(Environment newEnvironment, SessionFactory newSessionFactory, DockstoreWebserviceConfiguration.DiagnosticsConfig newConfig) {
```

Comment:
Just to be paranoid, is it worth making DiagnosticsHelper a singleton, and then ensuring this method can only be run once, like when you try to call `Thread.start()` twice on the same instance?

You don't have to implement this, and it is unlikely to happen the way our code is structured, but it did strike me when reviewing as the sort of thing you wouldn't want to accidentally instantiate multiple times.

Reply:
In general, the singleton pattern can make the code harder to test and less flexible, so IMHO, we should only use it when it's critical that the code in question only be instantiated once.  This isn't one of those cases, if we launch DiagnosticsHelper twice, we'll see the double logging and fix it real quick, but nothing will get corrupted or otherwise messed up.